base_revision/models/base_revision.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2013 Agile Business Group sagl (<http://www.agilebg.com>)
+# Copyright 2016 Serpent Consulting Services Pvt. Ltd.
+# Copyright 2018 Dreambits Technologies Pvt. Ltd. (<http://dreambits.in>)
+# Copyright 2020 Ecosoft Co., Ltd. (<http://ecosoft.co.th>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, api, fields, models
+
+
+class BaseRevision(models.AbstractModel):
+    _name = "base.revision"
+    _description = "Document Revision (abstract)"
+
+    @api.depends("old_revision_ids")
+    def _compute_has_old_revisions(self):
+        for rec in self:
+            rec.has_old_revisions = (
+                True if rec.with_context(active_test=False).old_revision_ids else False
+            )
+
+    current_revision_id = fields.Many2one(
+        comodel_name="base.revision",
+        string="Current revision",
+        readonly=True,
+        copy=True,
+    )
+    old_revision_ids = fields.One2many(
+        comodel_name="base.revision",
+        inverse_name="current_revision_id",
+        string="Old revisions",
+        readonly=True,
+        domain=["|", ("active", "=", False), ("active", "=", True)],
+        context={"active_test": False},
+    )
+    revision_number = fields.Integer(string="Revision", copy=False, default=0)
+    unrevisioned_name = fields.Char(
+        string="Original Reference", copy=True, readonly=True
+    )
+    active = fields.Boolean(default=True)
+    has_old_revisions = fields.Boolean(compute="_compute_has_old_revisions")
+    revision_count = fields.Integer(
+        compute="_compute_revision_count", string="Previous versions count"
+    )
+
+    @api.depends("old_revision_ids")
+    def _compute_revision_count(self):
+        res = self.with_context(active_test=False).read_group(
+            domain=[("current_revision_id", "in", self.ids)],
+            fields=["current_revision_id"],
+            groupby=["current_revision_id"],
+        )
+        revision_dict = {
+            x["current_revision_id"][0]: x["current_revision_id_count"] for x in res
+        }
+        for rec in self:
+            rec.revision_count = revision_dict.get(rec.id, 0)
+
+    _sql_constraints = [
+        (
+            "revision_unique",
+            "unique(unrevisioned_name, revision_number)",
+            "Reference and revision must be unique.",
+        )
+    ]
+
+    def copy(self, default=None):
+        default = default or {}
+        if "unrevisioned_name" not in default:
+            default["unrevisioned_name"] = False
+        revision_records = super().copy(default=default)
+        for rec in revision_records:
+            if rec.unrevisioned_name:
+                continue
```

Comment:
It can be simplified as
```suggestion
        for rec in revision_records.filtered(lambda line: not line.unrevisioned_name):
```